File: 2022/ts/day4.ts
import { assert } from "./deps.ts";
import { readLines } from "./utils.ts";

type Item = { start: number; end: number };
type Pair = [Item, Item];

function parseItem(token: string): Item {
  const [start, end] = token.split("-");
  return { start: parseInt(start), end: parseInt(end) };
}

export function parse(line: string): Pair {
  const tokens = line.split(",");
  assert(tokens.length >= 2);
  assert(tokens[0].length >= 2);
  assert(tokens[1].length >= 2);
  return [
    parseItem(tokens[0]),
    parseItem(tokens[1]),
  ];
}

export function contains(pair: Pair): boolean {
  const [item1, item2] = pair;
  return (item1.start >= item2.start && item1.end <= item2.end) ||
    (item2.start >= item1.start && item2.end <= item1.end);
}

export function run1(lines: string[]): number {
  return lines.filter((line) => line.length > 0).map((
    line,
  ) => contains(parse(line))).filter((contained) => contained).length;
}

function part1() {
  const part1 = run1(readLines("../inputs/input4"));
  console.log(`Part 1: ${part1}`);
}

part1();

export function overlap(pair: Pair): boolean {
  const [item1, item2] = pair;
  return (item1.start >= item2.start && item1.start <= item2.end ||
    item1.end >= item2.start && item1.end <= item2.end) ||
    (item2.start >= item1.start && item2.start <= item1.end ||
      item2.end >= item1.start && item2.end <= item1.end);
}

export function run2(lines: string[]): number {
  return lines.filter((line) => line.length > 0).map((
    line,
  ) => overlap(parse(line))).filter((contained) => contained).length;
}

function part2() {
  const part2 = run2(readLines("../inputs/input4"));
  console.log(`Part 2: ${part2}`);
}

part2();

File: 2022/python/day4.py
with open('../inputs/input4') as f:
	lines = f.readlines()
	part1 = 0
	part2 = 0

	for line in lines:
		items = []
		for tokens in line.strip().split(','):
			item = [int(token) for token in tokens.split('-')]
			items.append(item)
		range1 = set(range(items[0][0], items[0][1]+1))
		range2 = set(range(items[1][0], items[1][1]+1))
		if range1.issubset(range2) or range2.issubset(range1):
			part1 += 1
		if range1.intersection(range2):
			part2+= 1

	print('Part 1:', part1)
	print('Part 2:', part2)	